
---
title: "biocMultiAssay: exploring unified representations of multi-omics data"
author: "Vincent J. Carey, Levi Waldron et al."
date: "October 15, 2014"
output:
  pdf_document:
    toc: yes
  html_document:
    highlight: pygments
    number_sections: yes
    theme: united
    toc: yes
---

# Background

An increasingly common use case involves a set of
samples or patients who provide measurements on
multiple data types, such as gene expression,
genotype, miRNA abundance.  It will frequently
be the case that not all samples will contribute to
all assays, so some sparsity in the
set of 
samples $\times$ assays is expected.

# Basic demonstrative resources

Here are some very simple manipulations with
TCGA ovarian cancer data.  The data sizes are manageable
enough that the loadHub function is used to deserialize
all relevant data.

```{r chk,cache=TRUE}
library(biocMultiAssay)
ov = dir(system.file("extdata/tcga_ov", 
   package="biocMultiAssay"), full=TRUE)
drop = grep("pheno", ov)
if (length(drop)>0) {
  pdpath=ov[drop]
  ov=ov[-drop]
  }
#
# informal labels for constituents
#
tags = c("ov RNA-seq", "ov agilent", "ov mirna", "ov affy", "ov CNV gistic",
  "ov methy 450k")
#
# construct expt instances from ExpressionSets
#
elist = lapply(1:length(ov), function(x) new("expt", 
     serType="RData", assayPath=ov[x], tag=tags[x], sampleDataPath=ov[x]))
#
# populate an eHub, witha master phenotype data frame
#
ovhub = new("eHub", hub=elist, masterSampleData = get(load(pdpath)))
```
